### PR TITLE
Protecting access to out channel with mutex

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/flashlight#363.

@joesis This is an alternative to #11 that allows us to use just the s.out channel.

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -450,10 +451,23 @@ func (s *session) onSessionError(readErr error, writeErr error) {
 		c.close(false, readErr, writeErr)
 	}
 	s.closeOnce.Do(func() {
```
Comment (joesis) [inline session.go]: It's not quite clear to me why don't put this block into `s.Close()`. Once the underlying connection is closed in `s.Close()`, it's impossible to send anything.

Comment (joesis) [reply session.go]: It also reminds me that the buffered data you mentioned in https://github.com/getlantern/lampshade/pull/11/files/d420860019b840845797614bb292a3c0f2500be3#r152271731 can't be sent out anyway. TCP has the same behavior except for the rarely used "half-close".

Comment (oxtoacart) [reply session.go]: > It's not quite clear to me why don't put this block into s.Close(). Once the underlying connection is closed in s.Close(), it's impossible to send anything.

That's a good point, and likely a bug in the initial implementation. In Go, the behavior of a net.Conn is that any data written to it via `Write()` before calling `Close()` will be sent. I think we want the same for lampshade.

Comment (oxtoacart) [reply session.go]: Oh, I think I see why this hasn't been a problem in practice. The `net.Conn` that we use is actually the stream. The only cases in which the session gets closed directly are either if there's a network error or we hit our idle timeout.

**Pull request closed without merging**